torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -90,6 +90,66 @@ namespace nn {
 /// \endrst
 class SequentialImpl : public Cloneable<SequentialImpl> {
  public:
+  class NamedSubmodule {
```

Comment:
why do we need NamedSubmodule to be an internal type of Sequential as opposed to just some helper that creates OrderedDicts?

Reply:
`NamedSubmodule` doesn't have to be an internal type of Sequential, but it has to be a class type so that we can use its constructors to resolve the types of elements in braced-init lists such as `{"m1", M(1)}`. If `NamedSubmodule()` is just a pure helper function, it won't be able to understand the type of `{"m1", M(1)}`, since braced-init list doesn't inherently have a type.

Reply:
The comments here might be helpful: https://github.com/pytorch/pytorch/pull/17552/files#diff-ad62a46b3133f7412c5667d4c194359eR395